ref/utilsx1.c

```diff
@@ -0,0 +1,101 @@
+#include <string.h>
+
+#include "utils.h"
+#include "utilsx1.h"
+#include "params.h"
+#include "thash.h"
+#include "address.h"
+
+/*
+ * Generate the entire Merkle tree, computing the authentication path for
+ * leaf_idx, and the resulting root node using Merkle's TreeHash algorithm.
+ * Expects the layer and tree parts of the tree_addr to be set, as well as the
+ * tree type (i.e. SPX_ADDR_TYPE_HASHTREE or SPX_ADDR_TYPE_FORSTREE)
+ *
+ * This expecta tree_addr to be initialized to the addr structures for the 
+ * Merkle tree nodes
+ *
+ * Applies the offset idx_offset to indices before building addresses, so that
+ * it is possible to continue counting indices across trees.
+ *
+ * This works by using the standard Merkle tree building algorithm,
+ */
+void treehashx1(unsigned char *root, unsigned char *auth_path,
+                const unsigned char *sk_seed, const unsigned char *pub_seed,
+                uint32_t leaf_idx, uint32_t idx_offset,
+                uint32_t tree_height,
+                void (*gen_leaf)(
+                   unsigned char* /* Where to write the leaves */,
+                   const unsigned char* /* sk_seed */,
+                   const unsigned char* /* pub_seed */,
+                   uint32_t idx, void *info),
+                uint32_t tree_addr[8],
+                void *info)
+{
+    /* This is where we keep the intermediate nodes */
+    unsigned char stack[tree_height*SPX_N];
+
+    uint32_t idx;
+    uint32_t max_idx = (1 << tree_height) - 1;
+    for (idx = 0;; idx++) {
+        unsigned char current[2*SPX_N];   /* Current logical node is at */
+            /* index[SPX_N].  We do this to minimize the number of copies */
+            /* needed during a thash */
+        gen_leaf( &current[SPX_N], sk_seed, pub_seed, idx + idx_offset,
+                    info );
+
+        /* Now combine the freshly generated right node with previously */
+        /* generated left ones */
+        uint32_t internal_idx_offset = idx_offset;
+        uint32_t internal_idx = idx;
+        uint32_t internal_leaf = leaf_idx;
+        uint32_t h;     /* The height we are in the Merkle tree */
+        for (h=0;; h++, internal_idx >>= 1, internal_leaf >>= 1) {
+
+            /* Check if we hit the top of the tree */
+            if (h == tree_height) {
+                /* We hit the root; return it */
+                memcpy( root, &current[SPX_N], SPX_N );
+                return;
+            }
+            
+            /*
+             * Check if the node we have is a part of the
+             * authentication path; if it is, write it out
+             */
+            if ((internal_idx ^ internal_leaf) == 0x01) {
+                memcpy( &auth_path[ h * SPX_N ],
+                        &current[SPX_N],
+                        SPX_N );
+            }
+
+            /*
+             * Check if we're at a left child; if so, stop going up the stack
+             * Exception: if we've reached the end of the tree, keep on going
```

Comment:
I don't think we hit this exception.  Also remaining 4 nodes doesn't make sense.  Perhaps this is a remainder from the x4 version?